File: HeraldTest-Flutter2.0/ios/Runner/PayloadDataSupplier.swift
//
//  PayloadDataSupplier.swift
//  Runner
//
//  Created by Andi Grozdani on 12/8/21.
//

import Foundation
import Herald

class IllnessPayloadDataSupplier: PayloadDataSupplier {
    
    fileprivate var identifier: Int
    fileprivate var illnessStatusCode: Int
    fileprivate var date : Date
    fileprivate var phoneCode: Int
    
    init(identifier: Int, illnessStatusCode: Int, date: Date) {
        self.identifier = identifier
        self.illnessStatusCode = illnessStatusCode
        self.date = date
        self.phoneCode = 0
    }
    
    func setIdentifier(identifier: Int) {
        self.identifier = identifier
    }
    
    func setIllnessStatusCode(illnessStatusCode: Int) {
        self.illnessStatusCode = illnessStatusCode
    }
    
    func setDate(date: Date) {
        self.date = date
    }
    
    func setPhoneCode(phone: Int) {
        self.phoneCode = phone
    }
    
    func getIdentifier() -> Int{
        return identifier
    }
    
    func getIllnessStatusCode() -> Int{
        return illnessStatusCode
    }
    
    func getDate() -> Date{
        return date
    }
    
    func getPhoneCode() -> Int {
        return phoneCode
    }
    
    func payload(_ timestamp: PayloadTimestamp, device: Device?) -> PayloadData? {
        // Return our status as a series of Bytes in a Data object instance
        // Note: Should cache this if it doesn't change often as a member in order to save processing CPU and thus battery
        let result = PayloadData();
        result.append(UInt64(identifier)); // bytes 0-7
        result.append(toPayload()); // bytes 8-16
        return result;
    }
    
    func getIdentifierFromPayload(identifierPayload: PayloadData) -> Int {
        return Int(identifierPayload.data.uint64(0)!); // use first 8 bytes only
    }
    
    func getIllnessStatusFromPayload(statusPayload: PayloadData) -> StatusPayloadResults {
        return fromPayload(raw: statusPayload.data.advanced(by: 8)); // don't include first 8 bytes
    }
    
    func toPayload() -> PayloadData {
        let result = PayloadData()
        result.append(UInt64(self.date.timeIntervalSince1970))
        result.append(UInt8(self.illnessStatusCode))
        result.append(UInt8(self.phoneCode))
        return result
    }
    
    func fromPayload(raw: Data) -> StatusPayloadResults {
        
        let date : Date = Date(timeIntervalSince1970: Double(raw.uint64(0)!))
        let illnessStatusCode : Int = Int(raw.uint8(8)!)
        let phoneCode : Int = Int(raw.uint8(9)!)
        
        return StatusPayloadResults(date: date, illnessStatusCode: illnessStatusCode, phoneCode: phoneCode)
    }
}

class StatusPayloadResults {
    fileprivate var date : Date
    fileprivate var illnessStatusCode : Int
    fileprivate var phoneCode : Int
        
    init(date: Date, illnessStatusCode: Int, phoneCode: Int) {
        self.date = date
        self.illnessStatusCode = illnessStatusCode
        self.phoneCode = phoneCode
    }
    
    func getDate() -> Date {
        return date
    }
    
    func getIllnessStatusCode() -> Int {
        return illnessStatusCode
    }
    
    func getPhoneCode() -> Int {
        return phoneCode
    }
    
}

File: HeraldTest-Flutter2.0/android/app/src/main/java/com/example/herald_flutter/PayloadDataSupplier.java
package com.example.herald_flutter;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import io.heraldprox.herald.sensor.Device;
import io.heraldprox.herald.sensor.datatype.Data;
import io.heraldprox.herald.sensor.datatype.Date;
import io.heraldprox.herald.sensor.datatype.PayloadData;
import io.heraldprox.herald.sensor.datatype.PayloadTimestamp;
import io.heraldprox.herald.sensor.datatype.UInt64;
import io.heraldprox.herald.sensor.datatype.UInt8;
import io.heraldprox.herald.sensor.payload.DefaultPayloadDataSupplier;

public class PayloadDataSupplier extends DefaultPayloadDataSupplier {

    protected int identifier;
    protected int illnessStatusCode;
    protected Date date;

    void setIdentifier(int identifier) {
        this.identifier = identifier;
    }

    void setIllnessStatusCode(int illnessStatusCode) {
        this.illnessStatusCode = illnessStatusCode;
    }

    void setDate(Date date) {
        this.date = date;
    }

    public PayloadDataSupplier(int identifier, int illnessStatusCode, Date date) {
        this.identifier = identifier;
        this.illnessStatusCode = illnessStatusCode;
        this.date = date;
    }

    public PayloadData toPayload() {
        PayloadData result = new PayloadData();
        result.append(new UInt64(date.secondsSinceUnixEpoch())); // bytes at indexes 0 - 7
        result.append(new UInt8(illnessStatusCode)); // byte at index 8 (9th byte)
        return result;
    }

    @NonNull
    @Override
    public PayloadData payload(@NonNull PayloadTimestamp payloadTimestamp, @Nullable Device device) {
        PayloadData result = new PayloadData();
        result.append(new UInt64(identifier));
        result.append(toPayload());
        return result;
    }

    static int getIdentifierFromPayload(PayloadData identifierPayload) {
        return (int) identifierPayload.uint64(0).value; // use first 8 bytes only
    }

    static statusPayloadResults getIllnessStatusFromPayload(PayloadData statusPayload) {
        return fromPayload(statusPayload.subdata(8));
        // don't include first 8 bytes
    }

    public static statusPayloadResults fromPayload(Data raw) {

        Date date = new Date(raw.uint64(0).value);
        int illnessStatusCode = raw.uint8(8).value;

        return new statusPayloadResults(date, illnessStatusCode);
    }
}

final class statusPayloadResults {
    private final Date date;
    private final int illnessStatusCode;

    public statusPayloadResults(Date date, int illnessStatusCode) {
        this.date = date;
        this.illnessStatusCode = illnessStatusCode;
    }

    public Date getDate() {
        return date;
    }

    public int getIllnessStatusCode() {
        return illnessStatusCode;
    }
}